### PR TITLE
Diagnostic plot "displays"

### PR DESCRIPTION
**Pull request by kylebaron**
# Summary

This plot adds a feature set for creating multipanel displays of diagnostic plots. 

- `ETA` versus covariates
- `NPDE` versus covariates
- `CWRES` versus covariates
- `NPDE` goodness of fit plots
- `CWRES` goodness of fit plots

In general 
-  `CWRES` should behave just like `NPDE`. 
- When we plot versus covariates, they can be either categorical or continuous; the code picks that up
- There are `_list` versions of most paneled plots (where it makes sense) that get called under the hood; these `_list` plots return lists of plots or lists of lists of plots; basically everything up to the arrangement of plots on the grid
- Plots in the output lists are all named
- When you have this `_list` output, the user can arrange themselves via a `with()` method
- It's pretty hard to test this stuff; did my best to at lest exercise the code; see pdf attached here and I included the source code under `tests/testthat/rmd` to run again

## ETA
- `eta_covariate()` and `eta_covariate_list()`
- Default output is list of lists, with outer list the `ETAs` and the inner list are the covariates
- These plots can be transposed so that the outer set is the covariates and the inner set is the ETAs
- `eta_covariate()` an arrange by columns or rows
- `tag_levels` gets passed through to `patchwork` inside `eta_covariate()`

## NPDE - covariate

- `npde_covariate()` and `npde_covariate_list()`
- Similar setup to `eta_covariate()` but there is only one level rather than two
  - `npde_covariate()` returns a single plot, arranged
  - `npde_covariate()` returns a list of plots

## CWRES - coviariate

This should be pretty much identical to the setup we have for `npde_covariate`

## NPDE - diagnostics

- `npde_panel()` gives you a single panel with all the NPDE diagnositcs
- `npde_panel_list()` gives you a list of the component plots
- `npde_hist_q()` gives you just the NPDE histogram and q-q plot; no list option
- `npde_scatter()` gives you the scatter plots - NPDE versus TIME, TAD, PRED; no list option
- None of these plots have the possibility of transpose

## CWRES - diagnostics

This should be pretty much identical to the setup we have for `npde` diagnostics

[displays.pdf](https://github.com/metrumresearchgroup/pmplots/files/14069925/displays.pdf)

### CODE REVIEW

**Comment by KatherineKayMRG** [comment]
Also, for the npde plot (and maybe the cwres too?) would that have the option to be flexible? So for example, could we give the user the option of which plot to put in the middle? PRED, TIME or TAD? 

And what about the option to drop TIME or TAD? Then, if for example users drop TAD, it could naturally make a 2x2 grid with the option to swap PRED and TIME in one column


Screenshot 2023-08-23 at 9 04 36 AM
or 
Screenshot 2023-08-23 at 9 04 40 AM

**Comment by kylebaron** [comment]
@KatherineKayMRG  - what do you think about this? Return a list with sensible names that you can arrange?

``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(patchwork)
library(pmplots)
#> Loading required package: ggplot2

data <- pmplots_data_obs()
id <- dplyr::distinct(data, ID, .keep_all = TRUE)
x <- c("RF", "AAG", "SCR//Creatinine (mg/dL)", "CPc", "STUDYc", "AST")
y <- c("ETA1//ETA-CL", "ETA2//ETA-V2", "ETA3//ETA-KA")

l <- npde_panel_list(data, xname = "demothizone (ng/mL)")

with(l, (time + tad) / (q + hist) / pred)
#> `geom_smooth()` using formula = 'y ~ x'
#> `geom_smooth()` using formula = 'y ~ x'
#> `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.
#> `geom_smooth()` using formula = 'y ~ x'
```



``` r

wit
#> Error in eval(expr, envir, enclos): object 'wit' not found
```

<sup>Created on 2024-01-24 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by KatherineKayMRG** [comment]
I love this! 

Can you do something similar with other functions? Like the ETA functions (screenshot from MERGE) 

Screenshot 2024-01-25 at 10 43 02 AM

**Comment by kylebaron** [comment]
@KatherineKayMRG  - yeah, it'll be same idea for ETAs

``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(patchwork)
library(pmplots)
#> Loading required package: ggplot2

data <- pmplots_data_obs()
id <- dplyr::distinct(data, ID, .keep_all = TRUE)
x <- c("RF", "AAG", "SCR//Creatinine (mg/dL)", "CPc", "STUDYc", "AST")
y <- c("ETA1//ETA-CL", "ETA2//ETA-V2", "ETA3//ETA-KA")
```

# Standard panel

``` r
eta_covariate(id, x, y)[[1]]
#> `geom_smooth()` using formula = 'y ~ x'
#> `geom_smooth()` using formula = 'y ~ x'
#> `geom_smooth()` using formula = 'y ~ x'
```



# Roll your own

``` r
l <- eta_covariate_list(id, x, y)
with(l$ETA1, (RF+AAG)/STUDYc, tag_levels = "A")
#> `geom_smooth()` using formula = 'y ~ x'
```



<sup>Created on 2024-01-25 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by kylebaron** [comment]
``` r


# Standard panel - by row

``` r
x <- c("RF", "CPc", "STUDYc", "AAG", "SCR//Creatinine (mg/dL)",  "AST")
eta_covariate(id, x, y)[[1]]
#> `geom_smooth()` using formula = 'y ~ x'
#> `geom_smooth()` using formula = 'y ~ x'
#> `geom_smooth()` using formula = 'y ~ x'
```



# Standard panel - by column

``` r
x <- c("RF", "CPc", "STUDYc", "AAG", "SCR//Creatinine (mg/dL)",  "AST")
eta_covariate(id, x, y, byrow = FALSE)[[1]]
#> `geom_smooth()` using formula = 'y ~ x'
#> `geom_smooth()` using formula = 'y ~ x'
#> `geom_smooth()` using formula = 'y ~ x'
```



<sup>Created on 2024-01-25 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by KatherineKayMRG** [comment]
Very nice. Is there an option for the opposite? For example, if I want to look if there is a weight effect or a RF effect across all covs? 

Didn't know `eta_covariate` existed! What a great option when there are a limited number of covs!

**Comment by kylebaron** [comment]
>Is there an option for the opposite? For example, if I want to look if there is a weight effect or a RF effect across all covs?

Not at the moment; but agree we should be able to do that. I basically want to cover all of the code we currently have in place to do this sort of thing and make doing the "other" thing really easy.

**Comment by kylebaron** [comment]
>What a great option when there are a limited number of covs!

I like this too; it'd be great to start doing more of these panels and cut some figures / pages from the report. And limited covs is great situation to just dump everything out on a single page and move on.

**Comment by kylebaron** [comment]
@KatherineKayMRG 

Transposable display

``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(patchwork)
library(pmplots)
#> Loading required package: ggplot2

data <- pmplots_data_obs()
id <- dplyr::distinct(data, ID, .keep_all = TRUE)
x <- c("RF", "AAG", "SCR//Creatinine (mg/dL)", "CPc", "STUDYc", "AST")
y <- c("ETA1//ETA-CL", "ETA2//ETA-V2", "ETA3//ETA-KA")
```

``` r
eta_covariate(id, x, y)[[1]]
#> `geom_smooth()` using formula = 'y ~ x'
#> `geom_smooth()` using formula = 'y ~ x'
#> `geom_smooth()` using formula = 'y ~ x'
```



``` r
eta_covariate(id, x, y, transpose = TRUE, ncol = 1)[c(1,2)]
#> $RF
```



    #> 
    #> $AAG
    #> `geom_smooth()` using formula = 'y ~ x'
    #> `geom_smooth()` using formula = 'y ~ x'
    #> `geom_smooth()` using formula = 'y ~ x'



<sup>Created on 2024-01-25 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

File: R/displays.R [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+#' cont <- c("WT//Weight (kg)", "ALB//Albumin (mg/dL)")
+#' cat <- c("RF//Renal function", "CPc//Child-Pugh")
+#'
+#' cwres_covariate(data, x = c(cont, cat), tag_levels = "A")
+#' cwres_covariate_list(data, x = cont)
+#'
+#' @return
+#' `cwres_covariate()` returns single graphic of scatter plot diagnostics
+#' as a `patchwork` object that has been arranged using [pm_grid()] and
+#' `cwres_covariate()` returns the same component plots unarranged in a list.
```
Comment (kyleam) [inline R/displays.R:183]: `s/cwres_covariate/cwres_covariate_list/`

File: R/displays.R [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+#' Get a single graphic of `NPDE` or `CWRES` diagnostics versus `TIME`,
+#' `TAD` and `PRED`. Output can be in either long or compact format.
+#'
+#' @inheritParams npde_panel
+#' @param compact use `compact = TRUE` to get a more compact display; see
+#' **Examples**.
+#'
+#' @return
+#' A single graphic with three panels (`NPDE` or `CWRES` versus `TIME`, `TAD`
+#' and `PRED` as a `patchwork` object. The default behavior is to create a
```
Comment (kyleam) [inline R/displays.R:411]: Missing closing paren after PRED?

File: R/utils.R [hunk 3]
```diff
@@ -550,3 +550,10 @@ chunk_by_cols <- function(data, id_per_chunk, cols) {
 }
 
 force_digits <- function(x,digits) formatC(x,digits=digits,format = 'f')
+
+require_patchwork <- function() {
```
Comment (kyleam) [inline R/utils.R:554]: Fine as is, but `pm_grid` could now use `require_patchwork`.

File: R/displays.R [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+#' Get a single graphic of `ETA` versus continuous and / or categorical
+#' covariates (`eta_covariate()`) or the component plots as a list
+#' that can be arranged by the user (`eta_covariate_list()`).
+#'
+#' @param df a data frame to plot.
+#' @param x character `col//title` for covariates to plot on x-axis;
+#' see [col_label].
+#' @param y character `col//title` for ETAs to plot on y-axis; see [col_label].
+#' @param ncol passed to [pm_grid()].
+#' @param byrow passed through [pm_grid()].
```
[the remark this answers is not in the document]
Comment (kylebaron) [reply R/displays.R]: Fixed.

File: R/displays.R [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+#'
+#' eta_covariate(data, cont, y = etas)
+#' eta_covariate(data, cont, y = etas, transpose = TRUE)
+#'
+#' eta_covariate_list(data, x = cat, y = etas)
+#' eta_covariate_list(data, x = cat, y = etas, transpose = TRUE)
+#'
+#' @return
+#' `eta_covariate()` returns a list of plots arranged in graphics as a
+#' `patchwork` object using [pm_grid()] and `eta_covariate_list()` the same
```
[the remark this answers is not in the document]
Comment (kylebaron) [reply R/displays.R]: Fixed.

File: R/displays.R [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+#' plots, but unarranged as a list of lists.
+#'
+#' When `transpose` is `FALSE` (default), plots in a single graphic are grouped
+#' by the `ETA` the names of the list reflect that name (e.g., `ETA1`). When
```
[the remark this answers is not in the document]
Comment (kylebaron) [reply R/displays.R]: Fixed.

File: R/displays.R [hunk 7]
```diff
@@ -0,0 +1,473 @@
...
+#' cat <- c("RF//Renal function", "CPc//Child-Pugh")
+#'
+#' npde_covariate(data, x = c(cont, cat), tag_levels = "A")
+#' npde_covariate_list(data, x = cat)
+#'
+#'
+#' @return
+#' `npde_covariate()` returns single graphic of scatter plot diagnostics
+#' as a `patchwork` object that has been arranged using [pm_grid()] and
+#' `npde_covariate()` returns the same component plots unarranged in a list.
```
[the remark this answers is not in the document]
Comment (kylebaron) [reply R/displays.R]: Fixed.

File: R/displays.R (hunk 1 above)
Comment (kylebaron) [reply R/displays.R]: Fixed.

File: R/displays.R [hunk 8]
```diff
@@ -0,0 +1,473 @@
...
+#'
+#' @details
+#' The default value for `ncol` (1) means the two plots will be arranged in a
+#' single column, with the histogram on the top and quantile-quantile plot
+#' on the bottom. Using `ncol=2` will return an graphic with the plots
+#' side by side.
+#'
+#' @return
+#' A single graphic is returned, with a `NPDE`  or `CWRES` histogram and
+#' quantile-quantile plot arranged in a an object
```
[the remark this answers is not in the document]
Comment (kylebaron) [reply R/displays.R]: Fixed.

File: R/displays.R (hunk 2 above)
Comment (kylebaron) [reply R/displays.R]: Fixed.

File: R/utils.R (hunk 3 above)
Comment (kylebaron) [reply R/utils.R:552]: Done.

**Comment by kylebaron** [comment]
Thanks for reviewing, @kyleam ; I believe everything is addressed. 

On your suggestion, I did give `vdiffr` a try. It seems to work, I think? But I quickly ran into [this issue](https://github.com/r-lib/vdiffr/issues/126).  svg images are generally 800K to 1 MB, and I had 20 MB of these files after first pass. Although we're not subject to cran limitations, I pulled back on this b/c I'm not sure it's sustainable. What do you think? 

Although not a test, I did spin that demo doc into a vignette so at least functionality can be verified that way.

**Review by kyleam — APPROVED** [review]
I said:

> I haven't had a chance to interactively play around with it much, [...]

I have now, and it looks great.

I hit one snag:

``` r
library(pmplots)
#> Loading required package: ggplot2

data <- pmplots_data_obs()
l <- npde_panel_list(data)
with(l, (q+hist) / pred, tag_levels = "a")
#> Error in `ggplot_add()`:
#> ! Can't add `hist` to a <ggplot> object.
```

Loading the `patchwork` namespace fixes that.

One way to trigger the namespace to be loaded would be to make `with.pm_display()` call `require_patchwork()`, which I think it ought to do anyway given the `patchwork::plot_annotation()` call that's downstream of where the above error happens.

One other minor comment: the doc examples assign to `cat`.  I think it is common for users to paste this example code in their session to try things out, so it'd be nice if this used a different name (`cats`?) to avoid any confusion over the `cat()` function being clobbered in their session.

**Comment by kyleam** [comment]
> On your suggestion, I did give `vdiffr` a try. [...] svg images are generally 800K to 1 MB, and I had 20 MB of these files after first pass

re vdiffr: yeah, I think it makes sense to hold off given those sizes.

ggplot2 and patchwork are coming in much lighter:

```shell
# ggplot2

$ git ls-tree -lr v3.4.4 tests/testthat | sort -k4,4 -gr | \
  head | numfmt --field=4 --to=iec-i
100644 blob c6a2038aba44c5a65cf7ec34f8a0cce64643a9ba   104Ki tests/testthat/_snaps/coord-map/usa-mercator.svg
100644 blob 3d5fa0b240156383c70438a60efb7b4b120079bb   104Ki tests/testthat/_snaps/coord-map/coord-map-switched-scale-position.svg
100644 blob 17142781de60cb45c78ed218f2e61d2415611f89    90Ki tests/testthat/_snaps/geom-violin/grouping-on-x-and-fill-dodge-width-0-5.svg
100644 blob 56049d8ef60474a49f9083b1eebf524959593105    89Ki tests/testthat/_snaps/geom-violin/grouping-on-x-and-fill.svg
100644 blob 1494c6bd08fc55c884b86872c39308748d9ae75d    53Ki tests/testthat/_snaps/geom-violin/with-smaller-bandwidth-and-points.svg
100644 blob 1db22dd4418ee6feedae58e2223911f74add1862    52Ki tests/testthat/_snaps/geom-violin/with-tails-and-points.svg
100644 blob c86d890439ecec126825a6c8860e69c2888439a4    50Ki tests/testthat/_snaps/guides/align-facet-labels-facets-vertical.svg
100644 blob 84b81e2ba21e501c437e3ded4d9673a38ca2056c    49Ki tests/testthat/_snaps/guides/align-facet-labels-facets-horizontal.svg
100644 blob 8bec1ac1a6865986eb3840c63df315445a68e4ec    48Ki tests/testthat/_snaps/geom-violin/quantiles.svg
100644 blob 86a328e5b52836d3916313bb7b71c6a83fa4f934    48Ki tests/testthat/_snaps/geom-violin/dodging-and-coord-flip.svg

$ git ls-tree -lr v3.4.4 tests/testthat | grep snaps | grep svg | \
  awk '{ SUM += $4 } END { print SUM }' | numfmt --to=iec
3.3M

# patchwork

$ git ls-tree -lr v1.2.0 tests/testthat | sort -k4,4 -gr | \
  head | numfmt --field=4 --to=iec-i
100644 blob ccdad1aa3340c15c2532aac6ef3804a7ea43bd68    47Ki tests/testthat/_snaps/collect_axes/multi-cell-title-and-axis-collection.svg
100644 blob f92b91880180238cb7cfc16cf15dc1a83c0823f6    46Ki tests/testthat/_snaps/arithmetic/adding-to-all-subplots-patchwork-theme-bw.svg
100644 blob 75b2158f7976ad136ea204307129a87dd1a8b388    40Ki tests/testthat/_snaps/arithmetic/adding-to-all-on-level-patchwork-theme-bw.svg
100644 blob 897d826561ecb861224aabdd238cfcb8c5e37f17    34Ki tests/testthat/_snaps/arithmetic/complex-composition-p1-p2-p3-p4.svg
100644 blob 6a2aaefffe61f1e4c046e5b6228f4a15526ca50f    33Ki tests/testthat/_snaps/layout/setting-widths-as-units.svg
100644 blob 84f4baaa8f93f55820e882841b22c20f51a5f453    33Ki tests/testthat/_snaps/layout/setting-heights-as-units.svg
100644 blob 92e7fd0a626c368806bf19dbd677f24d0039ea83    33Ki tests/testthat/_snaps/arithmetic/pack-4-plots-p1-p2-p3-p4.svg
100644 blob a5bd2aa8071e7ad302cd5640cac1bd816d5cafec    33Ki tests/testthat/_snaps/layout/setting-heights.svg
100644 blob 9bc6de34ea29b4dcb2fd64aff9c2d983902aa4ba    33Ki tests/testthat/_snaps/layout/setting-widths.svg
100644 blob 51cabdcbcfb833b26646959f97b356e01fa4804d    33Ki tests/testthat/_snaps/layout/setting-nrow.svg

$ git ls-tree -lr v1.2.0  tests/testthat | grep snaps | grep svg | \
  awk '{ SUM += $4 } END { print SUM }' | numfmt --to=iec
847K
```

I guess it's probably a just a matter of how complicated the plots are (in terms of number of elements).

**Comment by kylebaron** [comment]
> I hit one snag:

Thanks; I updated that to `require_patchwork()`;  I was hoping that I tested that; but guessing the `patchwork` namespace was already loaded by the time I got to that test. I'll make a pass to make sure we're requiring `patchwork` anywhere in this series; maybe should just import it at some point going forward.

**Comment by kylebaron** [comment]
Actually ... I think it's the `with.pm_display()` method that really needs to require patchwork; once we get dispatched there, we are expecting to do some patchwork math that should always require that namespace to be loaded.

**Comment by kyleam** [comment]
> Actually ... I think it's the `with.pm_display()` method that really needs to require patchwork

Hmm, where were you originally thinking of adding the `require_patchwork` call?  `with.pm_display` was where I was suggesting that you call it (and what I tested locally).

**Comment by kylebaron** [comment]
@kyleam  - I ready too quickly and was thinking it was the upstream function that was lacking. Agree with your suggestion that modifying `with()` fixes this in the most comprehensive way. Thanks.

**Pull request merged**